### PR TITLE
P4OVS: Add documentation for linux networking

### PR DESCRIPTION
**Pull request by nupuruttarwar**
This patch adds documentation for linux networking

Title: Add documentation for linux networking
Change-type: Other
Signed-off-by: Nupur Uttarwar <nupur.uttarwar@intel.com>

### CODE REVIEW

File: p4proto/kctrl/switchapi/README.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ * **switch_api_l2_forward_create**: 
+  - Triggered by SwitchSAI create_fdb_entry API. Creates FDB handle based on type of information received (Tx or Rx). Tx FDB learn can be from tunnel interface, physical interface or VLAN interface.
+  - Table entry is created with platform dependent (PD) API (switch_pd_l2_tx_forward_table_entry for Tx entry and switch_pd_l2_rx_forward_table_entry for Rx entry).
+  - Once the table entry succeeds, handle along with api_l2_info is cached in l2 hashtable with mac address as the key.
+ * **switch_api_l2_forward_delete**:
+  - Triggered by SwitchSAI delete_fdb_entry API. Deletes the table entry using platform dependent API based on type of handle(Tx/Rx) and deletes the cache from Judy array
+
+## RIF <a name="rif"></a>
+ * **switch_api_rif_create**: Handle is allocated and initialized for handle type SWITCH_HANDLE_TYPE_RIF and the port ID for which this function is triggered, is determined
+using platforming dependent API switch_pd_to_get_port_id and saved in rif_info structure.
```
Comment (Namrata-intel) [inline p4proto/kctrl/switchapi/README.md]: platform

**Review by Namrata-intel — APPROVED** [review]

**Review by namratalimaye — APPROVED** [review]
Looks good.

File: p4proto/kctrl/switchapi/README.md (hunk 1 above)
Comment (nupuruttarwar) [reply p4proto/kctrl/switchapi/README.md]: Fixed the typo

**Review by namratalimaye — APPROVED** [review]
Looks good

File: p4proto/p4src/linux_networking/README.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+3. [Create P4 artifacts](#create_p4_artifacts)
+3. [Limitations](#limitations)
+4. [Steps to create topology](#steps)
+
+## Overview <a name="overview"></a>
+This README describes a step by step procedure to run Linux networking scenario
+
+## Topology <a name="topology"></a>
+<p align="center">
+  <img alt="Linux Networking Topology" src="p4proto/p4src/linux_networking/topology.png">
```
Comment (n-sandeep) [inline p4proto/p4src/linux_networking/README.md]: looks like this image is broken can you please check

File: p4proto/kctrl/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+P4OVS kernel control plane listens to kernel events using Switchlink library which provides
+a netlink listener and uses Switch Abstraction Interfcae(SAI) layer to program the P4 compliant
+target by interacting with a target agnostic frontend layer TDI (Table Driven Interface) through
+Switchapi layer. p4proto/kctrl directory contains the code to manage kernel interface
+
+## Table of Contents
+
+- [Switchlink](/p4proto/kctrl/switchlink/README.md)
+- [SwitchSAI](/p4proto/kctrl/switchsai/README.md)
+- [Switchapi](p4proto/kctrl/switchapi/README.md)
```
Comment (n-sandeep) [inline p4proto/kctrl/README.md]: this hyper link didnt work for me, can you please check

File: p4proto/p4src/linux_networking/lnw.conf [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    "instance": 0,
+    "p4_devices": [
+        {
+            "device-id": 0,
+            "eal-args": "dummy -n 4 -c 3",
+            "mempools": [
+                {
+                    "name": "MEMPOOL0",
+                    "buffer_size": 2304,
+                    "pool_size": 1024,
```
Comment (n-sandeep) [inline p4proto/p4src/linux_networking/lnw.conf]: this conf file looks to be old, please check the latest one.
We have modified this pool_size to 4096,
better to keep it same as the default conf file we copy to /usr/share/stratum

File: p4proto/p4src/linux_networking/README.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                   phy-port0         - 2
+                   TAP1              - 3
+                   phy-port1         - 4
+                   TAP2              - 5
+                   TAP0              - 6
+                   TAP3              - 7
+```
+ - **Rule for: Any tx control packet from VM1(TDP 0), pipeline should add a VLAN tag 1 and send it to TAP0 port(TDP 6)**
+  ```
+  ovs-p4ctl add-entry br0 linux_networking_control  .handle_tx_control_pkts_table "istd.input_port=0,action=linux_networking_control.push_vlan_fwd(6,1)"
```
Comment (n-sandeep) [inline p4proto/p4src/linux_networking/README.md]: please remove space between for the table name
linux_networking_control & handle_tx_control_pkts_table

**Comment by n-sandeep** [comment]
are we going to have a separate README.md, topology.PNG for ECMP?

File: p4proto/p4src/linux_networking/README.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+* Bind the devices to DPDK using dpdk-devbind.py script
+```
+    cd $SDE_INSTALL/bin
+    ./dpdk-devbind.py --bind=vfio-pci <pci_bdf> eg: ./dpdk-devbind.py --bind=vfio-pci 0000:18:00.0
+```
+ *Note*: pci_bdf can be obtained using lspci command. Check if device is binded correctly using ./dpdk-devbind.py -s (Refer to the section "Network devices using DPDK-compatible driver")
+
+#### 2. Export the environment variables and start running the OVS
+```
+    source $SDE_INSTALL <$DEPS_INSTALL>
```
Comment (n-sandeep) [inline p4proto/p4src/linux_networking/README.md]: correct this command.
source p4ovs_env_setup.sh $SDE_INSTALL <$DEPS_INSTALL>

File: p4proto/kctrl/switchapi/README.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+ * **switch_pd_ecmp_hash_table_entry**
+```
+    Table name: LNW_ECMP_HASH_TABLE
+    Key: LNW_ECMP_HASH_TABLE_KEY_HOST_INFO_TX_EXTENDED_FLEX_0
+    Action: LNW_ECMP_HASH_TABLE_ACTION_SET_NEXTHOP_ID with parameters nexthop_id
+
+#### Tunnel Platform Dependent API
+
+ * **switch_pd_tunnel_entry**
```
Comment (n-sandeep) [inline p4proto/kctrl/switchapi/README.md:153]: This is not aligned with previous hyperlinks in the document
eg: switch_pd_ecmp_hash_table_entry

File: p4proto/kctrl/switchapi/README.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+```
+    Table name: LNW_VXLAN_ENCAP_MOD_TABLE
+    Key: LNW_VXLAN_ENCAP_MOD_TABLE_KEY_VENDORMETA_MOD_DATA_PTR
+    Action: LNW_VXLAN_ENCAP_MOD_TABLE_ACTION_VXLAN_ENCAP with
+    parameters src_addr, dst_addr, VNI
+```
+
+ * **switch_pd_tunnel_term_entry**
```
Comment (n-sandeep) [inline p4proto/kctrl/switchapi/README.md:161]: same as above

File: p4proto/kctrl/switchsai/README.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+## Tunnel <a name="tunnel"></a>
+
+* **create_tunnel:** Takes tunnel attributes as input, creates tunnel entry in SwitchAPI with switch_api_tunnel_create, and returns tunnel handler to the caller.
+* **remove_tunnel:** Takes tunnel handler as an input, deletes tunnel entry in SwitchAPI with switch_api_tunnel_delete.
+* **create_tunnel_term_table_entry:** Takes tunnel terminate attributes as input, creates tunnel term entry in SwitchAPI with switch_api_tunnel_term_create and returns the handler to the caller.
+* **remove_tunnel_term_table_entry:** Takes tunnel termination handler as an input, deletes tunnel termination entry in SwitchAPI with switch_api_tunnel_term_delete.
+
+## FDB <a name="fdb"></a>
+
+FDB entry is populated from 2 sources, one is from SwitchLink→SwitchSAI→SwitchAPI and other is from OVS Mac Learning→ SwitchAPI. Here we talk only about FDB entry from SwitchLink.
```
Comment (n-sandeep) [inline p4proto/kctrl/switchsai/README.md:74]: somehow these are shown in a different way while we view it on browser.
https://github.com/ipdk-io/ovs/blob/linux_networking_documentation/p4proto/kctrl/switchapi/README.md#fdb-

File: p4proto/kctrl/switchapi/README.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+
+## Overview <a name="overview"></a>
+The Switchapi library is an abstraction layer that exposes higher level API's used to interact with target backend to program the data plane described in the linux_networking.p4.
+Modules supported in switchapi are FDB, Nexthop, Neighbor, ECMP, Route, VRF, RMAC, Tunnel.
+
+For each module, as part of initialization, device context is set, hashtable is initialized and a unique handle is allocated and initialized based on the type of module.
+
+## FDB <a name="fdb"></a>
+ * During initialization, device context is set, hashtable is initialized and handle is allocated for both Tx (SWITCH_HANDLE_TYPE_L2_FWD_TX) and Rx (SWITCH_HANDLE_TYPE_L2_FWD_RX) and updated in device context
```
Comment (n-sandeep) [inline p4proto/kctrl/switchapi/README.md:40]: somehow these are shown in a different way while we view it on browser.
https://github.com/ipdk-io/ovs/blob/linux_networking_documentation/p4proto/kctrl/switchapi/README.md#fdb-

File: p4proto/kctrl/switchapi/README.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+#### FDB platform dependent API:
+ * **switch_pd_l2_tx_forward_table_entry**
+   *  For FDB learns from vlan interface
+```
+    Table name: LNW_L2_FWD_TX_TABLE
+    Key: LNW_L2_FWD_TX_TABLE_KEY_DST_MAC
+    Action: LNW_L2_FWD_TX_TABLE_ACTION_L2_FWD with parameters port id of vlan interface
+```
+   * For FDB learns from physical interface
```
Comment (n-sandeep) [inline p4proto/kctrl/switchapi/README.md:92]: these doesnt show as sub sections in, it shows in the same level as switch_pd_l2_tx_forward_table_entry
https://github.com/ipdk-io/ovs/blob/linux_networking_documentation/p4proto/kctrl/switchapi/README.md#fdb-platform-dependent-api

File: p4proto/kctrl/switchapi/README.md (hunk 7 above)
Comment (nupuruttarwar) [reply p4proto/kctrl/switchapi/README.md:153]: Fixed the alignment

**Comment by nupuruttarwar** [comment]
> are we going to have a separate README.md, topology.PNG for ECMP?
yes

File: p4proto/kctrl/switchsai/README.md (hunk 9 above)
Comment (n-sandeep) [inline p4proto/kctrl/switchsai/README.md]: With the latest change, FDB entry populated from ovs also takes SwitchSAI path.
OVS Mac Learning→SwitchSAI→SwitchAPI

we can also remove "Here we talk only about FDB entry from SwitchLink"

File: p4proto/kctrl/switchapi/README.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+ * **switch_api_router_mac_group_delete** - Iterates through all entries in rmac list amd delete the rmac entry and handle associated
+
+## Neighbor <a name="neighbor"></a>
+ * **switch_api_neighbor_create**: Handle is allocated and initialized for handle type SWITCH_HANDLE_TYPE_NEIGHBOR. Retrieves the nhop_handle from api_neighbor_info attribute passed by SwitchSAI. If nhop handle is valid, create the handle for neighbor and update the nexthop table and neighbor mod table using appropriate platform dependent (PD) API's
+ * **switch_api_neighbor_delete**: Retrieve the neighbor_info from neighbor handle and call the platform dependent (PD) API's to delete table entry from target backend and delete the handle
+
+## Nexthop <a name="nexthop"></a>
+ * **switch_api_nhop_create**: Handle is allocated and initialized for handle type SWITCH_HANDLE_TYPE_NHOP. Check if the nhop handle exists, if not, create one and get the nhop key based on nhop type. Once the handle is successfully created, cache the nhop info with key as nhop key in the hashtable and return nhop handle to SwitchSAI
+ * *Note*:
```
Comment (n-sandeep) [inline p4proto/kctrl/switchapi/README.md]: this looks little unaligned, please check

File: Makefile.am [hunk 13]
```diff
@@ -122,6 +122,12 @@ EXTRA_DIST = \
 	OVS_WITH_P4_ARCH.rst \
 	CONTRIBUTING_P4OVS.rst \
 	OVS-WITH-P4-BUILD-README \
+	p4proto/kctrl/README.md \
```
Comment (n-sandeep) [inline Makefile.am]: Why are we adding these files here?
shouldn't these files be added as part of EXTRA_DIST in p4proto/automake.mk?

File: p4proto/kctrl/switchapi/README.md (hunk 12 above)
Comment (nupuruttarwar) [reply p4proto/kctrl/switchapi/README.md]: Alignment is fine here: 
image
Please let me know if you still see any issue

File: Makefile.am (hunk 13 above)
Comment (nupuruttarwar) [reply Makefile.am]: Since these are README's I have added them in global Makefile, let me know if any concerns with this

**Review by n-sandeep — APPROVED** [review]
looks good to me

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-35-9/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/0a24032efad37502573ef1eb05d8aa89157982cd

**Pull request merged**